services/brig/src/Brig/User/Auth.hs

```diff
@@ -125,7 +122,35 @@ login (PasswordLogin li pw label _) typ = do
     AuthSuspended -> throwE LoginSuspended
     AuthEphemeral -> throwE LoginEphemeral
     AuthPendingInvitation -> throwE LoginPendingActivation
+  verifyCode code uid
   newAccess @ZAuth.User @ZAuth.Access uid typ label
+  where
+    verifyCode :: Maybe Code.Value -> UserId -> ExceptT LoginError (AppIO r) ()
+    verifyCode mbCode u = do
+      (mbEmail, mbTeamId) <- getEmailAndTeamId u
+      featureStatus <- lift $ do
+        mbStatus <- Intra.getTeamSndFactorPasswordChallenge `traverse` mbTeamId
+        pure $ maybe (Public.tfwoapsStatus Public.defaultTeamFeatureSndFactorPasswordChallengeStatus) Public.tfwoStatus mbStatus
+      case featureStatus of
+        Public.TeamFeatureDisabled -> pure ()
+        Public.TeamFeatureEnabled -> do
+          case mbCode of
+            Nothing -> loginFailedWith LoginCodeRequired u
+            Just c ->
+              case mbEmail of
+                Nothing -> loginFailedNoEmail u
+                Just email -> do
+                  key <- Code.genKey <$> (Code.mk6DigitGen $ Code.ForEmail email)
+                  codeValid <- isJust <$> Code.verify key Code.AccountLogin c
+                  if codeValid
+                    then pure ()
+                    else loginFailedWith LoginNoPendingCode u
```

Comment:
aesthetics: can this multiple-intended case/if chain be flattened out a little for readability? E.g. through the use of `ifNothing` or similar, or through the use of a `case (x, y, z) of ...` pattern ?